### ISSUE TITLE
Validate tokens and macros on load

### ISSUE THREAD

**Issue by Phergus** [RPTools/maptool#939]
**Is your feature request related to a problem? Please describe.**
A recent bug caused macros on copied tokens to have UUIDs identical to the original macros causing a few issues.  Similar situations have arisen in the past as well.

**Describe the solution you'd like**
Each token and macro should be checked during campaign/map/token load to check that they have unique UUIDs and if they don't then create a new UUID for it.  Changes to be logged with sufficient info for a user to locate the changed token or macro.

This should help catch and prevent similar issues in the future.  Would also provide a place in the code for future sanity/validity checks to be added in the future if needed.

**Comment by Merudo** [comment]
I suggest we also check for broken token IDs that were generated from older MapTool versions.

**Comment by JamzTheMan** [comment]
Re older GUID for tokens, yea, can be a simple check that if it starts with a numeral regenerate, new UUID always starts with an alpha character now.

**Comment by Phergus** [comment]
Do people commonly stash token IDs in token properties or in macros themselves for use later?   I think it should still be done but was wondering if there is going to be complaints where someone has stashed a bunch of UUIDs in a token property or note field and suddenly some of them aren't valid anymore.  I suspect/hope that isn't a common practice but...

**Comment by JamzTheMan** [comment]
I don't think we've officially ever said UUID's are guaranteed to be persistent across saves, only unique. We have had a long standing issue where token ID's regenerated on server start for a long while (did we fix that yet?)

I suppose we could log to chat if we need to change an ID so they would know if they did?

**Comment by Merudo** [comment]
I fixed the token ID reset on server start a few versions ago.

UUID will be unique in 1.5.8, in 1.5.7 and below the UUID were duplicated when copying tokens.

**Comment by Phergus** [comment]
Fair enough.  I called out the logging of changes in the original post.  I was thinking to the actual log file but could also go to chat.

**Comment by JamzTheMan** [comment]
@Merudo k, I thought you fixed it.

Re logging, ya, def to log file but as you pointed out, the user may want to know an I'd was force changed and who reads the log unless you are coding?

Also, as we develop new releases, if we break it again in the future it should be easier to see/catch if it errors to chat?

**Comment by Phergus** [comment]
True.  More likely to catch during development that way.
